### ISSUE TITLE
DSpace should have an Audit Trail of object changes

### ISSUE THREAD

**Issue by tdonohue** [DSpace/DSpace#8824]
**Is your feature request related to a problem? Please describe.**
DSpace should have an out-of-the-box way to track common changes to various objects: who made the change, when was the change made, and what was the change.

For example, a way to track these sort of conceptual changes
* Item [UUID] was created by [user] on [date/time]
* Item [UUID] was moved by [user] on [date/time] to Collection [UUID]
* Item [UUID] was updated by [user] on [date/time]  (possibly with information about updates, e.g. metadata fields added/removed/updated)
* Item [UUID] was withdrawn/reinstated/deleted by [user] on [date/time]
* Collection [UUID] was created/updated/deleted...
* Community [UUID] was created/updated/deleted...
* Bitstream [UUID] was created/updated/deleted...

**Describe the solution you'd like**
Ideally, DSpace should track these changes automatically (and likely internally) in a location where the audit trail can be easily reviewed/searched (especially for individual objects).  This might mean the audit trail needs to be in the database and linked to the object it relates to. (It might be stored as private/internal metadata, or could be a separate database table altogether.)

Only users with Admin rights on the object should be able to see the audit trail of that object.  The audit trail should be accessible from the object page at a minimal (so that you can see all changes to that object). But, a global audit search/browse option may need to also be implemented to find information about object *deletion* (as the object no longer exists in that scenario).

Performance must also be a consideration, as the audit trail could grow large for large systems.  This may imply that the audit trail needs tools to configure what information is tracked (based on needs of institution) or cleanup/delete older entries (e.g. auto-remove all audit entries older than 1-2 years).

**Describe alternatives or workarounds you've considered**
Currently, the only workaround for DSpace 7 or below is to use the existing `dspace.log` and `dc.description.provenance` metadata fields as a very basic "audit trail".
* `dc.description.provenance` only exists for Items. It only tracks some very basic actions on the Item (e.g. audit of creation information, withdraw/reinstate information, workflow approval information)
* `dspace.log` provides INFO logs for many audit trail like tasks (object updates/creation/deletion).  But, not all tasks of interest may be logged.  Searching older logs can also be complex as you must understand the log format and also possibly manage/index older logs in some manner.

**Additional context**
This topic has come up on mailing lists before ([one example](https://groups.google.com/g/dspace-community/c/W_8NF80yFGQ/m/cSz7hc94AgAJ)).  But, also, more recently came up in the discussion of improving logging in #8777

**Comment by steph-ieffam** [comment]
@tdonohue we are actively working on this feature. Please notice in order to implement this feature the following PR/issue should be merged https://github.com/DSpace/DSpace/pull/10684 https://github.com/DSpace/DSpace/issues/10683

**Comment by steph-ieffam** [comment]
this should be planned for version **10** since no new features are accepted on version 9

**Comment by paulo-graca** [comment]
@tdonohue and @steph-ieffam 

We would like to share our perspective on the work being proposed as a solution to this issue.

Based on our experience migrating services to DSpace 7, as well as the feedback we have received from our users, we can already state that the user experience feedback leaves us in a difficult position, unable to provide a positive response. We attempted to mitigate some aspects of the user experience by developing a custom theme. However, theming cannot address the underlying issues caused by architectural and implementation decisions.

At the same time, in version 7.6.1, where some performance issues were not yet resolved (though later they were), we have faced severe performance problems. These vary depending on the complexity of each repository installation (as not all repositories are alike), and we have been forced to find workarounds. Here are some examples of contributions which, although not yet part of DSpace core, significantly improve user experience:

- https://github.com/DSpace/DSpace/pull/11110
- https://github.com/DSpace/DSpace/pull/11333

In addition to the complaints received about performance and user experience, we have also observed that the institutions we support are increasingly seeking alternative solutions that allow them to manage the entire scientific process. It was only one in the past, but the number has slowly growing. We recognize there are more appropriate forums to discuss these broader concerns, and this ticket may not be the correct place.

We had already expressed this before: this is a required feature. For example, we currently use one of our own
https://github.com/fccn/rcaap-dspace-additions/blob/4efe12028417a8d730b6ff472fc1d13d297818bb/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V5.0_2021.12.21__RCAAP_before_DS7.sql#L5-L82

to do perform basic audit tasks, such as knowing when an item was created/deleted. While not as thoroughly detailed as the proposed solution, it has been sufficient, along with logs, for our tracking needs.

That said, the proposed work raises us several concerns:

### User Interface / User Experience

The proposed interface has usability issues from our point of view. When tested with an experienced repository manager, the user requires an explanation of what's the purpose, yet still struggled to understand what was being displayed. Clearly, the current interface is not intended for this user profile.

Even if it were intended for a system administrator profile, such a role has not yet been clearly defined in DSpace. It may make more sense to first define and design this profile before adding functionality for it. Otherwise, we risk accumulating ad-hoc features that later require significant integration work.

Even technically experienced system administrators will also struggled to interpret what's in the interface.

We believe a more user-friendly approach could be to extend the visualization of the item (or the DSpaceObject) itself, where, depending on permission levels, with the activity change displayed (displayed in a specific context). Deleted items would, of course, required to be retained and searchable.

### Performance Concerns

From our experience with DSpace 7, there are still many rough edges. For instance, after migrating a service from DSpace 5 to 7, we noticed persistent robotic activity trying to fetch non-existent pages (https://github.com/DSpace/dspace-angular/issues/4421). DSpace did not handle these requests efficiently, leading to server overload and requiring extreme countermeasures, like blocking traffic.

This is just one example of the type of problems we face, which consume time we would otherwise dedicate to new upgrades. Therefore, we are very concerned — perhaps even afraid — that this new functionality, as currently designed, could further impact performance.

While we understand the feature can be turned on/off, in our case we will still require some form of auditing (or continue using our current audit tables if necessary: https://github.com/fccn/rcaap-dspace-additions/blob/4efe12028417a8d730b6ff472fc1d13d297818bb/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V5.0_2021.12.21__RCAAP_before_DS7.sql#L5-L82).

Our specific concerns include:
- If a user with database access deletes a record directly, the change will not be logged.
- Migration between servers would make it difficult to persist the audit data. Export/import from Solr is possible but adds administrative overhead that may be overlooked.
- Because the solution relies on the Event mechanism, execution is immediate. For intensive tasks (imports/ingests), this introduces additional overhead.
- Storage is in Solr. For typical installations, what is the expected overhead? Will additional infrastructure be required? For comparison: our current solution has been in production environment since 2014, and in a repository with 108k records, we already have 270k activity changes (nearly 3x more than the number of items), and this is only tracking items and handles (we excluded metadata as it would generate excessive operations logs).
- For small/simple installations, what will be the additional load of enabling this solution?

### Final Consideration

In all honesty, we question whether a dedicated user interface for this functionality is needed at all. For us, logs with the audit activity would be sufficient - unless a Sysadmin profile is formally defined and taken into consideration.

**Comment by steph-ieffam** [comment]
@paulo-graca and @tdonohue 
@paulo-graca Thank you for sharing your perspective on this issue. I would like to address some of the points you raised and provide additional context based on our experience implementing this feature.

First, regarding community feedback, I'm not entirely clear about the negative feedbacks mentioned. The provided Pull Request actually aims to introduce a new functionality that many institutions have specifically requested: an easily accessible way to verify what changes have been made to an item (particularly metadata changes). While logs do partially contain this information, they don't integrate well with the UI, making it difficult for repository managers to access this data efficiently.

I'm somewhat surprised by the concern about UI complexity. We have multiple clients actively using this feature, and their repository managers haven't reported difficulties understanding or using this section. There are significantly more complex functionalities in DSpace, both in terms of UI configuration and workflow comprehension (e.g., replication with AIP, COAR Notify and others). The Audit feature simply displays rows in a table showing all events generated for that item and related objects. Perhaps adding a disclaimer explaining Subject ID and Object ID concepts could be beneficial, but beyond that, the functionality seems fairly straightforward.

Regarding the suggestion to display different UIs based on user experience level, this would require evaluating user groups at minimum. We would need to distinguish between administrators who are repository managers and those who are administrators but lack technical understanding, which seems impractical. Additionally, preserving deleted items in DSpace would be a breaking change that I'm not sure should be introduced alongside this feature. Repository administrators already have the option to flag items as withdrawn rather than completely removing them from the system. Making deleted item management similar to withdrawn items seems redundant and of limited utility.

On performance concerns: if the audit feature is not enabled, it has absolutely no impact on performance. With the property set to false, someone could even remove the consumer entirely to have one less in the list. We've also experienced issues with scraping bots, but I don't see a clear connection between those problems and this feature.

Addressing your specific concerns:

- **Regarding direct database modifications**: indeed, these wouldn't be tracked by any audit solution. However, we don't believe this level of modification needs tracking - users making such changes are typically system administrators who should know what they're doing and their changes are usually tracked at a higher level in the istitution. At that level only a system dump would help in understanding what has happened. The audit is designed to inform repository managers or administrators about changes made through DSpace (UI, endpoints, or scripts) for various purposes (identifying script errors, tracking incorrect metadata changes, or determining who made specific modifications or even tracking odd bitstream changes).
- **On migration issues**: this is something that was mentioned in a community meeting, but I consider it a non-issue. Solr is not fundamentally different from a PostgreSQL database - it's a NoSQL database (mostly used as a search engine) that can be backed up just like any other database https://solr.apache.org/guide/solr/9_0/deployment-guide/backup-restore.html. Solr information should be treated exactly like any other database information, as is already done with statistics. SQL databases are strongly discouraged for this particular use case because, given the volume of data to manage, a single operation could take several minutes to complete in PostgreSQL while in SOLR a couple of seconds are required for the most complex operations (faceting for instance).
- **On event-based implementation**: we're leveraging DSpace's existing event system. The execution is immediate, just as it would be when calling a dedicated java method somewhere else in the code or inserting rows in an SQL database. Bulk edits or mass imports don't occur frequently, and any new feature implementation would naturally introduce some delay wheter it's this Audit feature or something else. Again, enabling this feature is up to the repository manager.
- **On Solr scaling**: we have several installations successfully using this feature. One institution has approximately 200-220K items, some migrated from DSpace CRIS 5 and others from external systems, with periodic import/update scripts running. All of this operates with audit enabled (tracking all the changes since the creation of item) without any special requirements for Solr. It's not significantly different from managing statistics, and I've seen statistics cores with 200 million records not requiring any special architecture. This ,again, proves the great reliabilty of SOLR
- **Finally**, the event/consumer approach offers flexibility. We can modify the consumer configuration in dspace.cfg to decide what events to handle if we're not interested in intercepting all events for all DSpace objects. We could also skip changes to items still in workflow or workspace, further reducing the data load sent to Solr by limiting it to changes made after archiving.

I hope this clarifies our approach and addresses your concerns. I look forward to further discuss on how we can refine this feature to meet the community's needs providing as much data as possible.

Stefano

**Comment by paulo-graca** [comment]
I may not have expressed myself in the best way. And, by not having introduced myself, that may not have helped to provide a clearer context for our concerns and the reason why they arise here. Broadly speaking, I work for an NREN, integrated within a national science funder. Our work is specifically related to the open science movement and the facilitation of its adoption. To this end, we support higher education institutions, hospitals, and laboratories through services and infrastructures.
Since the first day of the service go-live, back in 2008, we have adopted DSpace as the repository solution. Since then, we have grown together. Our service has already been used as the birthplace of some of the most relevant DSpace features, along side with the Minho's University, in features such as: XOAI, Request a Copy, support for OpenAIRE 4 guidelines, among others. I should also add that we are actively involved with the community: we participated in the discussion that led to the end of the dual JSPUI + XMLUI solution; we promoted, participated in, and supported the adoption of the concept of Entities in DSpace.

I would also add that our specific work is to provide support to the more than 180 institutions we interact with, including scientific journals, offering services in the form of SaaS. This diversity, at least on our scale, allows us to have comparative terms between services and systems. When I refer here to issues related to user experience, or even performance, I base it both on the various complaints reported to us and on what we actually observe in the services we provide.

Not wanting to deviate too far from the main topic of the ticket, I will just leave a small example with this issue: https://github.com/DSpace/dspace-angular/issues/4582 – was created because the person responsible for managing the repository, who works with it on a daily basis (we support, but don't use it in our daily basis), complained that they could not tell who holds a given task. Being in a position of responsibility, they are unable to manage the people involved or follow the status of the work, because the installed solution does not provide that information. They use a paper and a pen to keep track of the workflow... This is just a small example.

In our service, we have the privilege of being in direct and permanent contact with the community we support, i.e., with those who actually use DSpace in their daily lives. When I raised the issues of UX or performance, I was referring precisely to the feedback we receive.

In my humble opinion, and in line with the way we like to work, it is essential to test in advance the design of any solution involving user interaction. That is exactly what I did with the screens provided in the PR: I used them to test the flow with a repository manager. The feedback I shared came from that interaction, since the manager had never been exposed to that functionality. I believe that, after some time, the user will eventually get used to and adapt to the functionality. However, UX best practices point to the opposite – the experience should be intuitive from the very beginning.

As for performance, which has been our biggest struggle in recent times, I have nothing further to add. Nor did I feel either more confident or less apprehensive after what was presented.

I do not, in any way, wish to force a solution that is not the best for DSpace. I would only like to add that, with a simple SQL instruction, we can keep a record, in the database, of all INSERT and DELETE operations with the Item table. The same applies to operations carried out directly in the database by users with database access privileges:
```sql
BEGIN;

-- CREATE FUNCTION "process_item_audit()" ----------------------
CREATE OR REPLACE FUNCTION process_item_audit()
 RETURNS trigger
 LANGUAGE plpgsql
AS $function$
    BEGIN
        IF (TG_OP = 'DELETE') THEN
            INSERT INTO item_audit_ds7 SELECT 'D', now(), user, OLD.*;
            RETURN OLD;
        ELSIF (TG_OP = 'INSERT') THEN
            INSERT INTO item_audit_ds7 SELECT 'I', now(), user, NEW.*;
            RETURN NEW;
        END IF;
        RETURN NULL;
    END;
$function$;
-- -------------------------------------------------------------

COMMIT;
```

**Comment by steph-ieffam** [comment]
Thank you @paulo-graca , for introducing yourself, but it wasn’t necessary. I value everyone’s opinions in the community, and although I see that you work with many institutions, the experience of 4Science is no less important.

The reason I feel cautious about recommending a database-based solution for an Audit functionality is that while it may work well for some cases (generally smaller institutions), as the number of items in a repository grows, the amount of audit information to store also increases, bringing several challenges.

From what I understand, one of your “worst” use cases has just 270K audit rows recorded, which is still a relatively low number and therefore easily manageable within a database table.

If we imagine deploying such a solution on DSpace, we would have to start with an Audit feature that does not record all the data institutions might want to track. Moreover, we should not think just about today, but what would happen if the feature is expanded to track other information? For example, metadata, additional info on group changes, or anything else an institution may need to track. In that case, the rows in the DB table would increase exponentially, easily reaching numbers in the millions.

So, although a DB-side approach may seem to work fine, it imposes important limits on future development of this functionality without requiring a future re-engineering. From our experience with repositories of various sizes, we have noticed that DSpace currently has significant limits in managing large volumes of information. Without going into every detail of each problem (which is not the purpose of this issue), I feel the preferred approaches should always be those that are easily scalable and enable both small and large institutions to operate without difficulties.

Unfortunately, experience teaches that when SQL database tables exceed millions rows, problems start to arise that were not previously evident, and analyzing database operations these become clear.

Also, from a Crawling perspective, considering the high number of requests that more and more institutions face, which could saturate the database connection pool or simply add more load, I feel discouraged about any solution that might add huge extra load on the database since it seems to be one of the most affected resources by these problems. Locking operations caused by audit writes on the DB could generate many more problems than those we already have. SOLR, instead, seems not to suffer from these issues or at least much less so.

Leaving aside the technical aspects where I wanted to highlight critical points and positives, I believe some DSpace functionalities could certainly be improved and made clearer. However, I do not think we should discard new features or simplify them just because users find them difficult to understand (perhaps simply because they did not receive proper training).

This argument applies to any existing DSpace feature. Group, user, and workflow management generally require specific training or at least documentation reading. I do not expect many DSpace features to be immediately intuitive given the complexity of the product and its functionalities. Although I believe that reading a table showing event, date, and user is already quite easy to understand and that's not only my humble opinion but is also shared by different repository managers. 

That said, we are always open to making DSpace better and improving it constantly. If something is unclear on the UI side, we would like to receive feedback on what could be improved instead of simply dismissing the feature as too complex for a repository manager.

Moreover, the solution with PostgreSQL-side trigger, while technically interesting, would mean moving all the logic to the database side, and I am not sure this is the best approach.
Having said that, I want to thank you for this interesting discussion, which I believe contributes positively to the community.

**Comment by paulo-graca** [comment]
First of all, @steph-ieffam I would like to thank you for all the time that was dedicated to providing a response, and also to thank you - if I have not done so yet - for the initiative of proposing a concrete solution to address this issue. Conceptually, I do not disagree with what has been said. Relational databases have both positive and negative aspects, as do non-relational databases. They serve different purposes. We have had both positive and negative experiences with both types when subjected to extreme activity or robotic activity. And I would be more than happy to share data.

I would just like to stress that I was not trying to propose what I shared here as the solution. I only wanted to share what we are using in our systems and which, together with the DSpace and system logs, has served us for supporting auditing activities. If I was not clear enough before, I would like to make that clear now. 

I will not add anything further to what I have already explained. However, I believe it would be helpful for others if we had access to concrete data to reassure us that enabling this new feature will not introduce system overhead or usability issues (I would suggest consulting DCAT).

**Comment by mwoodiupui** [comment]
Regarding RDBMS vs. Solr:  audit data are a log.  Why would we want to use any sort of DBMS, whether PostgreSQL or Solr?  Audit recording should be simple, rugged and fast, suggesting a sequential file.

There are a number of good log analysis tools already.  Is this really the sort of thing that should be built into DSpace?

**Comment by tdonohue** [comment]
@paulo-graca and @steph-ieffam :  Just getting to this discussion.  While I think you've done a good job of getting back to a cordial discussion, I'd like to make sure that we ***please keep discussions on topic within GitHub tickets***.   While I welcome feedback on other areas of DSpace and how they perform (or UX issues you've run into), those need to go into their own separate tickets (or into tickets that may already exist on that topic, if any).  Otherwise, I worry that feedback will be "lost" as soon as this ticket is closed.

I'd like to also stress that (as you both know) we are a volunteer-based development group.  So, solutions that are brought forward by any one person / team / group _will often not be perfect_. That's OK, in my opinion, as long as it's a step in the "right direction" and something we all feel we can build upon and improve upon.  That includes even if the UI is less than perfect.  I feel our goal is to find a level that we all generally agree is "good enough" / "moving in the right direction".  From there, we can continue to improve upon it via ongoing collaboration and contributions of other volunteers.

In the case of this proposed audit trail feature from 4Science, I've heard two main points that several Committers ([at least in the meeting on Sept 25](https://wiki.lyrasis.org/display/DSPACE/2025-09-25+DSpace+Developers+Meeting)) felt where necessary to answer or address:

1. **Concerns about storage in Solr only:** The main concern I've heard from several is that the storage solution of "only storing in Solr" is not ideal (and I'm one that also feels this way).  I think it's perfectly OK to use Solr for *indexing*, but I'd prefer some form of "permanent" storage layer...or at least the plans for such a solution.  One of the biggest frustrations that I continue to hear about our DSpace upgrade process is the upgrade process for Usage Statistics (which are also stored only in Solr), as people do not like how they have to export those statistics and reimport in order to reindex them.  I think we should learn from that frustration, and brainstorm if there's anything we can do that is "better" for an Audit Trail.  Otherwise, I worry we'll compound that frustration by forcing people to do this same "export and reimport" process for both Statistics and Audit Trail.
    * This doesn't mean that we have to store the audit trail in the database. But, it could mean we should consider whether it's possible to "recreate" the audit trail by indexing your log files?  Or maybe we need separate audit log files which can be kept and reindexed easily? (It could even be an optional flag to decide if you want to keep audit logs alongside Solr or not.)  These ideas are similar to what @mwoodiupui expressed above.
2. **Lack of Seach/Filter in Global Audit Trail:** As we also heard, there's some concerns about how to locate *deletions* in the Audit Trail.  The current approach expects users to mostly use the audit trail at the object level.  That works well enough, except when it comes to deletions.  In that scenario, if you wanted to understand who deleted an object, you'd have to use the global audit trail, which has no search/filter options.  This would be painful, as you'd have to continually page through audit events until you can locate the exact deletion you are looking for.  If that deletion happened a few days ago or a week ago, it'd be even more difficult to find.

**Comment by steph-ieffam** [comment]
Hi all.

@mwoodiupui regarding the decision to store audit logs in SOLR: this choice was made because it gives administrators direct access to the modified audit data. Not everyone is comfortable working with raw log files, and as already mentioned, SOLR seemed like a good compromise.

As for the two points @tdonohue  highlighted again:
Yes, facets and filters are currently missing in the global section, and at the moment the audit is mainly accessible from the pages of the individual DSpaceObjects that have been modified/created. This is not integrated into the current PR, but it would be possible to create additional Pull Requests to improve this functionality.

On the topic of backup, I believe the concern comes from the bias of considering SOLR mainly as a faster-access copy of what is already stored in the database (which is indeed true for the SOLR search core), with the database itself being the ultimate source of data. However, as I pointed out, SOLR includes built-in backup features (https://solr.apache.org/guide/solr/9_0/deployment-guide/backup-restore.html
) that can be used.
Migration challenges were mostly an issue with older SOLR versions. In practice, this is not very different from what happens today when migrating a database like PostgreSQL and with this new feature there seems to be no need for exporting using CSV or odd formats.

I think we could discuss again this at the next developer meeting, so we can make it clear that there are more efficient alternatives than the classic old-style CSV export to backup and restore SOLR.
To be honest, the idea of using “log files” for audit and then reindexing them feels quite similar to exporting and reimporting a CSV into SOLR—but maybe I simply misunderstood what was being proposed. CSV seems to me as a more machine-readable log format

**Comment by tdonohue** [comment]
@steph-ieffam : Just to clarify the feedback I've received in the past about User Statistics being only in Solr.  The frustration that I've heard is around having to run this process:
```
[dspace]/bin/dspace solr-export-statistics -i statistics
```
to first export your statistics.  Then you have to run this to reimport / reindex:
```
[dspace]/bin/dspace solr-import-statistics -i statistics
```

I'm told by sites with larger amounts of Statistics that this process is **painfully slow**.  We're lucky it doesn't have to be run in *every DSpace upgrade*. But, it was required to in the upgrade to DSpace 7, and it's required anytime a Solr upgrade requires a full reindex of the cores (or anytime our Solr schemas change for that core).  See https://wiki.lyrasis.org/display/DSDOC7x/Upgrading+DSpace where this was required for the backend upgrade in steps 2b and 10a

So, my primary goal is to avoid duplicating this existing frustration to an Audit Trail feature, as that may make the DSpace upgrade process even more frustrating for larger sites.

That said, if we can find a way to somehow reindex in-place, or in a less time-consuming manner, then I'd be OK with it. 

My suggestion of audit logs is just to potentially *simplify* the reindex process during upgrades.  It could be more easily automated into the `./dspace index-discovery` script or a similar script, without needing to migrate that data separately.  While it can help with backups as well, I'm more concerned about finding ways to ease our upgrade processes, and avoid adding more extra steps specific to reindexing your audit trail.

**Comment by mwoodiupui** [comment]
The argument against Solr as primary store is not bias -- the Solr maintainers will tell you it's designed to be a fast cache (plus fancy text analysis) for data that can be reloaded from elsewhere.  Solr "backup" is a snapshot of the Lucene core files, suitable only for disaster recovery, and won't load into a newer version.  Every full-function export/import for Solr that I know of is a custom-written dump/restore tool.  And a reload from external source is required for version upgrades and schema changes, as noted above.

(Anything I say about audit storage also applies to statistics storage.  They are both huge, ever-growing, long-lived time series of data that cannot be regenerated but must from time to time be reindexed.  There are already good tools for searching, selecting, analyzing and visualizing both types of data.)

**Comment by steph-ieffam** [comment]
Thanks for sharing your feedback
Especially in the latest versions, Solr is an search engine that can easily serve as a primary data store for certain scenarios.
Some production systems use Solr (or Elasticsearch, which is very similar) as the sole data store for logs or large-scale analytics. It would not be acceptable if it were “just a cache layer” and this is currently also happening for our statistics core.
The original claim holds some truth when looking at older versions of Solr. It is also possible to perform an export in CSV text format, which is essentially equivalent to using log files to track all changes when and if neened only that in this case integration with the DSpace UI is feasible.

With proper changes in the schemas, it is possible to introduce new fields without tearing down the index. Moreover, as long as you avoid changing the tokenizer or the type of an existing field, reindexing is not required. Reindexing becomes necessary only when the Solr schema is drastically changed and old data is no longer compatible with the new schema. However, this does not happen when schema changes simply involve adding new fields rather than altering existing ones.

Regarding the reports of slow scripts, I can certainly agree with the community. However, in my humble opinion, a script that takes too long should be reviewed and optimized rather than reconsidering the underlying technology. Leveraging multithreaded approaches could be an idea. That said, if schema changes are carefully planned, such an operation should not be necessary at all.

**Comment by tdonohue** [comment]
@steph-ieffam : I've added this topic to our next Developers Meeting agenda. https://wiki.lyrasis.org/display/DSPACE/2025-10-09+DSpace+Developers+Meeting

I'm not sure I fully agree with your analysis of current Solr. The [current Solr's documentation on reindexing](https://solr.apache.org/guide/solr/latest/indexing-guide/reindexing.html) notes that the **["best approach" to reindexing](https://solr.apache.org/guide/solr/latest/indexing-guide/reindexing.html#delete-all-documents) is to *delete all documents* in your Solr core, and then "reindexing by re-running the original index process"**.  

It also details that sometimes a reindex is required in order to [upgrade Solr](https://solr.apache.org/guide/solr/latest/indexing-guide/reindexing.html#upgrades) (depending on whether your current index is still compatible with the new version of Lucene)  The example it gives is that you can successfully run Solr 8 with indexes built from Solr 7.  But, if you later want to upgrade to Solr 9, that version will be *incompatible with indexes from Solr 7* which means you'd have to do a full reindex.

(In DSpace's case, DSpace 8 used Solr 8 and DSpace 9 uses Solr 9.  So, if you built indexes in DSpace 8, you'll have to rebuild them as soon as we move to Solr 10.  This just means that we need to plan for a reality where semi-regular Solr reindexing will need to occur.)

The challenge with following the Solr documentation is that this Audit Trail solution and current Usage Statistics have no way to "re-run" the original index process unless the content either exists somewhere else (like in a log file) or is first exported (like the current `solr-export-statistics` script).  

So, it's unclear to me how we will reindex this Audit Trail information when we are next required to reindex Solr? Is the plan to use the existing `solr-export-statistics` script to export the entire Audit Trail and reindex?  Is there some other plan?  

I think we need a clear plan for how this Audit Trail will be reindexed before we can move forward with this feature.

Regardless, we can talk more in next week's Developers Meeting as necessary.

**Comment by tdonohue** [comment]
As discussed in [today's Developers Meeting](https://wiki.lyrasis.org/display/DSPACE/2025-10-09+DSpace+Developers+Meeting), the main concern is around **using Solr as the primary storage for the Audit Trail**.

It was clear based on discussions (and the chat) that there is some disagreement around the strategy. 
* All seem to agree that we should index the Audit Trail into Solr, as that'll make it easier to view/search in the User Interface.  
* No one seems to be advocating for using RDBMS or database.
* However, there's several who are advocating for storing (or backing up automatically) Audit Trail data on the filesystem, perhaps as logs or perhaps as more structured data (JSON/CSV).
    * There was even mention of whether this Audit Trail data could somehow be stored in the assetstore, and managed alongside other "archived" files.

The main concerns with keeping the primary storage in Solr relate to "pain points" with dealing with [DSpace's Usage Statistics](https://wiki.lyrasis.org/display/DSDOC9x/SOLR+Statistics) (which also use Solr as primary storage).  The main pain point expressed is that the process for reindexing Usage Statistics using `solr-export-statistics` and `solr-import-statistics` scripts can take a very long time (with reports of it taking many days or even a week).

In the end, there seems to be two main options for moving this feature forward:
1. **Keep Solr as primary data storage while adding a "performance friendly" way to easily reindex:**  This would require creating a script similar to `solr-export-statistics` and `solr-import-statistics` to allow for easily reindexing/reimporting Audit Trail information.
    * _Ideally_, we'd even find a way to perhaps create a generic `solr-export` and `solr-import` scripts which provide better performance and can be used for *both* Audit Trail and Statistics (as that'd provide consistency) 
2. **Or, ensure Audit Trail information is also stored in file(s):**  This would involve ensuring that Audit Trail information is stored also in (preferably structured) text file(s).  This might be CSV or JSON file(s), or well structured log files.
    * These file(s) should be optimized for quick re-indexing, allowing you to recreate or reindex your Audit Trail into Solr during upgrades.
    * These files(s) could be more easily backed up, preserving your audit trail information, without having to backup Solr.
    * We could consider providing a flag to turn off text file storage (for sites who don't want to duplicate the data).  But, in that scenario, those sites would still need a way to reindex their Audit Trail in Solr (similar to the first bullet above).

Either of these two approaches is acceptable to move forward.  The primary goal is to ensure we are not duplicating the frustrations that people have with storing Usage Statistics in Solr (as primary storage), and either approach will work around that problem.

**Comment by paulo-graca** [comment]
Regarding Solr as the primary solution option, I would like to highlight that we are currently migrating repositories between servers and versions. One of the required steps in this process is converting Solr statistics to the new format using `solr-upgrade-statistics-6x`.

We have more than 50 million entries, and the process takes about 10 hours to handle batches of 10 million records. A simple calculation shows that it would take around 5 days for this step to complete for a single repository.

```
           9,050,000  Processed... (    46 sec; 9:01:05; DB cache:  1,021/ 128,199; Docs: 10,000)
           9,060,000  Processed... (    34 sec; 9:01:39; DB cache:  1,021/ 128,462; Docs: 10,000)
           9,070,000  Processed... (    37 sec; 9:02:17; DB cache:  1,021/ 128,694; Docs: 10,000)
           9,080,000  Processed... (    42 sec; 9:02:59; DB cache:  1,021/ 128,877; Docs: 10,000)
           9,090,000  Processed... (    36 sec; 9:03:36; DB cache:  1,025/ 129,064; Docs: 10,000)
           9,100,000  Processed... (    37 sec; 9:04:13; DB cache:  1,025/ 129,261; Docs: 10,000)
           9,110,000  Processed... (    40 sec; 9:04:54; DB cache:  1,025/ 129,407; Docs: 10,000)
           9,120,000  Processed... (    34 sec; 9:05:28; DB cache:  1,025/ 129,586; Docs: 10,000)
           9,130,000  Processed... (    41 sec; 9:06:10; DB cache:  1,025/ 129,683; Docs: 10,000)
           9,140,000  Processed... (    37 sec; 9:06:48; DB cache:  1,025/ 129,862; Docs: 10,000)
           9,150,000  Processed... (    38 sec; 9:07:26; DB cache:  1,025/ 129,988; Docs: 10,000)
```

**Comment by steph-ieffam** [comment]
@tdonohue we followed up on your suggestions, and in seeking a simple solution for everyone, we decided it would be best to implement a Java script to manage SOLR cores in a fully generic way. This means that the solution presented in this PR is capable of handling (both importing and exporting) data from any SOLR core, including both audit and statistics cores.

By default, the script runs with a single thread, but the proposed solution fully supports parallel threading, improving the SOLR management script performance. Thread configuration is customizable, allowing users with limited server performance to process data normally (generally single thread), while giving those with more powerful servers greater flexibility to optimize processing.

The solution has been tested on both the statistics and audit cores to provide a fast and reliable way to export data. As I mentioned during the last developers’ meeting, it is absolutely possible to export and import data from SOLR without having to wait weeks for the process to complete, that long waiting time usually depends on slowness introduced by un-needed java operations.
 I also reviewed the script you mentioned as being slow and noticed that it performs certain operations that should be avoided, acceptable for small datasets, but not for institutions with large volumes of data.
However, that script is quite legacy and appears tightly coupled to the statistics core, so I preferred to propose a new version that uses the new scripting framework, avoiding direct removal of the old statistics script which we could later deprecate once this new solution is validated for that core as well.

I would like to emphasize that, being aware of the importance of code reuse, I've chosen to define a unified solution capable of managing audit and other cores in a general way. That said, I’d like to keep the focus of this thread on the audit functionality, as repeatedly mentioned, since refactoring multiple DSpace features at once doesn’t seem appropriate within such a specific issue and the main goal of the script should have been allowing to export and import (so reindexing is done) for the Audit Core.

With that said, I’d like to share with you the basic operation of the export/import script while waiting for the creation of the documentation space for version 10, where I’ll be able to upload the complete audit documentation.

```
./dspace solr-core-management 
    -m <mode:{export|import}> 
    -c <core:{audit|statistics|...}> 
    -d <directory> 
    [-f <format:{csv|json}>] 
    [-t <threads:integer>=1] 
    [-s <start-date:yyyy-MM-dd>] 
    [-e <end-date:yyyy-MM-dd>] 
    [-i <increment:{WEEK|MONTH|YEAR}>] 
    [-h]
```

- -m, --mode → Execution mode. Must be export or import.
- -c, --core → SOLR core name. Source core for export or target core for import.
- -d, --directory → Directory path where files are written (export) or read (import).
- -f, --format → File format: csv or json. Default: csv.
- -t, --threads → Number of worker threads to use. Default: 1.
- -s, --start-date → Start date for export in yyyy-MM-dd format. Optional.
- -e, --end-date → End date for export in yyyy-MM-dd format. Optional.
- -i, --increment → Time interval for splitting export: WEEK, MONTH, or YEAR. Default: MONTH.
- -h, --help → Shows usage information and exits.

I tested out the export script on the audit and statistics core and specifically for the statistic core it managed to export with a single thread mode (the server was a shared test server running on low CPU and memory - 2 Core 6GB RAM) about **110 millions** of statistics data in less than **4 hours**.
I welcome anyone to run the same tests and also test out the import flow. For the import flow I've also run a couple of tests using the same cores (audit & statistics)
@tdonohue please let me know what you think about the script and the feature itself.

**Comment by tdonohue** [comment]
@steph-ieffam : Based what you've described of the [proposed script above](https://github.com/DSpace/DSpace/issues/8824#issuecomment-3397474134), that approach seems reasonable to me.   

I do agree that you are **not** expected to solve/fix all the existing User Statistics issues.  In the above comments, I was simply using User Statistics as an _example_ of a feature that works similarly to your Audit Trail approach. 

So, I also agree that it's perfectly OK to leave the existing `solr-export-statistics` and `solr-import-statistics` scripts as they are.  But, hopefully, as you noted, we will find that the new `solr-core-management` script will be an accurate replacement -- if so, we can remove the old scripts in a future PR.

Overall, this all seems like a good approach to me.

**Comment by steph-ieffam** [comment]
@tdonohue I just wanted to share the documentation about the Audit feature but I'm currently in doubt about where the documentation page should be placed. I see there's no version 10.0 documentation an since 10.0 is the latest unreleased version I though it could be a good idea to place this documentation in the 10.0 doc.
Should I add this feature’s documentation anywhere else?
I also want to highlight that I have no permission to create new pages even though I can edit pages all around the docs.
Let me know, thanks a lot

**Comment by tdonohue** [comment]
@steph-ieffam : I've just created a new space for the 10.0 documentation at https://wiki.lyrasis.org/display/DSDOC10x   Currently, this wiki page is an exact copy of the 9.x documentation.  But, you are welcome to start adding new DSpace 10.0 documentation to that area.
